### PR TITLE
Will/popup

### PR DESCRIPTION
**Pull request by WillHHH**

### CODE REVIEW

File: src/components/Popup/index.scss [hunk 1]
```diff
@@ -0,0 +1,39 @@
+@import '../../styles/colors.scss';
+
+$baseClass: 'vant-popup';
+$baseContainerClass: 'vant-popup__container';
+
+.#{$baseContainerClass} {
+  visibility: hidden;
+  position: absolute;
+  background-color: transparent;
+  z-index: 5;
```
Comment (mxdi9i7) [inline src/components/Popup/index.scss:10]: 5?

File: src/components/Popup/index.scss [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+    &__isActive {
+      background-color: $default;
+      display: flex;
+      justify-content: center;
+      align-items: center;
+    }
+
+    &__center {
+      height: 200px;
```
Comment (mxdi9i7) [inline src/components/Popup/index.scss:35]: save these into variables.scss

File: src/components/Popup/index.scss [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  visibility: hidden;
+  position: absolute;
+  background-color: transparent;
+  z-index: 5;
+	transition: 1s;
+
+
+  &__isActive {
+    visibility: visible;
+    background-color: rgba($color: #000000, $alpha: 0.4);
```
Comment (mxdi9i7) [inline src/components/Popup/index.scss:16]: color

Comment (mxdi9i7) [reply src/components/Popup/index.scss:16]: alpha saved into variables

File: src/components/Popup/index.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  component: Popup
+};
+
+export const PopupType = () => {
+  const [centerPopup, setCenterPopup] = useState(false);
+  const [centerPopup2, setCenterPopup2] = useState(false);
+
+  return (
+    <div className='container'>
+      <button
```
Comment (mxdi9i7) [inline src/components/Popup/index.stories.tsx:16]: use our own button

File: src/components/Popup/index.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,44 @@
+import React, { useState } from 'react';
+import Popup from './';
+import '../../styles/stories.scss';
+
+export default {
+  title: 'Popup',
+  component: Popup
+};
+
+export const PopupType = () => {
```
Comment (mxdi9i7) [inline src/components/Popup/index.stories.tsx:10]: give these types a more clear name

File: src/components/Popup/index.tsx [hunk 6]
```diff
@@ -0,0 +1,90 @@
+import React, { useRef, useState, useEffect } from 'react';
+
+import classnames from '../../utils/classNames';
+import Icon from '../Icons';
+
+import './index.scss';
+
+export interface IProps {
```
Comment (mxdi9i7) [inline src/components/Popup/index.tsx:8]: put this in a types.ts file

File: src/components/Popup/index.tsx [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+  isActive: boolean;
+  borderRadius?: string;
+  text?: string;
+  position?: 'center' | 'top' | 'bottom' | 'left' | 'right';
```
Comment (mxdi9i7) [inline src/components/Popup/index.tsx:12]: create a type for position

File: src/utils/classNames.ts [hunk 8]
```diff
@@ -1,10 +1,11 @@
 const classnames = (baseClass: string, names: object[]): string => {
   const normalizedNames: string[] = [];
+  const haveValue: string[] = ['type', 'position'];
```
Comment (mxdi9i7) [inline src/utils/classNames.ts:3]: name this array something like: specialKeys

**Comment by z695245771** [comment]
image
image
Looks bad on windows dude

File: src/components/Popup/types.ts [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+  children?: string;
+  closeable?: boolean;
+  closeIcon?: [string, string];
+  closeIconPosition?: object;
+  setActive: Function;
+  click?: Function;
+}
+
+export type PopupTypes = 'center' | 'top' | 'bottom' | 'left' | 'right';
+export type TextAlign = 'left' | 'right' | 'center';
```
Comment (mxdi9i7) [inline src/components/Popup/types.ts:20]: We already have this align type

File: src/components/Popup/index.tsx [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+
+  const isInclude: Function = (data: string) => {
+    return popupProps.className.includes(data);
+  };
+
+  if (borderRadius)
+    Object.assign(popupProps, {
+      style: {
+        ...popupProps.style,
+        borderTopLeftRadius:
```
Comment (mxdi9i7) [inline src/components/Popup/index.tsx:96]: this logic needs to be simplified, also you need to use class name here

File: src/components/Popup/index.tsx [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+  text,
+  content,
+  borderRadius,
+  type = 'center',
+  color,
+  size,
+  isActive,
+  setActive,
+  click,
+  closeIcon = ['cross', '20px'],
```
Comment (mxdi9i7) [inline src/components/Popup/index.tsx:22]: why is this an array??

File: src/utils/classNames.ts [hunk 12]
```diff
@@ -1,10 +1,11 @@
 const classnames = (baseClass: string, names: object[]): string => {
   const normalizedNames: string[] = [];
+  const specialKeys: string[] = ['type'];
   names.forEach((name: object) => {
     if (Object.prototype.toString.call(name) === '[object Object]') {
       const key = Object.keys(name)[0];
       const value = name[key];
-      if (key === 'type') {
+      if (specialKeys.includes(key)) {
```
Comment (ssysm) [inline src/utils/classNames.ts:8]: Why use `.includes`? Can't you use `.indexOf`? It's much faster.

**Pull request merged**